ck4up.1
.TH ck4up 1 "Sep 29 2026" "ck4up 1.5" ""
.SH NAME
ck4up \- Check for Updates, a utility to monitor web pages for updates 
.SH SYNOPSIS
\fBck4up [options] [<regexp> ...]\fP
.SH DESCRIPTION
\fBck4up\fP is a small command-line utility, written in ruby, primarily 
intended for CRUX ports maintainers to watch http and ftp sites for 
updates, but may also be useful for others.

\fBck4up\fP scans through a configuration file, fetches the listed URLs 
from the web, computes the md5sum of the page, and compares the value with
one stored in a gdbm database. If both differ a message will be written
to the standard output.

To improve performance, \fBck4up\fP is multi-threaded. A configurable
number of threads fetches pages in parallel. The default setting is a
maximum of 20 threads and can be changed at the top of the script
(variable Threads_max).

Regular expressions can be applied to each page in the event that \fBck4up\fP
produces repetitive diff messages. For example, if daily snapshots and
a web counter are on the same page.

To simplify maintenance of the configuration file, a kind of 
macro-expansion can be used.

.SH OPTIONS
.TP
.B -d, --debug
Run ck4up in debug mode. All fetched pages will be written to the
standard output. This is especially useful when combined with a
regexp to limit the number of pages to process.
.TP
.B -k, --keep
Keep md5 values, don't update the database.  
.TP
.B -v, --verbose
Verbose mode. Print a message also for unchanged pages.
.TP
.B -h, --help
Print a short help message and exit.
.TP
.B -p, --parseonly
Parse configuration file, expand macros and print lines matched by <regexp>. 
Do not fetch any remote data.
.TP
.B -c, --cleandb
Removes unused entries from the ck4up database.
.TP
.B -x, --exact
Regexp arguments are used as exact matches for the names in
the config file. Example: if you have definitions for both "st"
and "vttest", the argument "st" normally matches both. Using
-x avoids this (only "st" is matched).
.TP
.B -w, --warn
If none of the regexp arguments match any lines in the config
file, print a warning to that effect on stderr, and exit with
non-zero status.
.TP
.B -f <file>, --config <file>
Use alternative configuration from <file> instead of the default 
\fI ~/.ck4up/ck4up.conf\fP. 
ck4up creates a database with the same full pathname but the extension
replaced by .dbm.
.TP
.B <regexp>
Process only configuration-lines matching <regexp>. If more than one
expression is given, all of them are or'ed to build a regular expression.

.SH CONFIGURATION

The configuration file for ck4up contains lines with macro
definitions and lines defining URLs to check for updates.
Blank lines and lines beginning with a pound-sign, which are 
considered comments, are ignored. Spaces at the beginning of
definition-lines are not allowed.
.br
The default location for the \fBck4up\fP configuration file 
is 
\fI~/.ck4up/ck4up.conf\fP.


.B <name>  <type>  <url>  [<regexp>]

Lines starting with a valid character in the range [a-zA-Z] are URL lines, 
defining the web-pages you would like to check for updates.
<name> is a symbolic name for the URL, md5 is the only valid 
entry for <type> at the moment, <url> is the URL of the 
web-site, <regexp> is an optionally extended regular expression.

Don't use spaces in regexps, but \\s instead. Also, if you're
using parentheses for alternation, use non-capturing ones:

This is OK: (?:foo|bar)

Don't use this: (foo|bar)

Examples:

.br
apache  md5  http://www.apache.de/dist/httpd/    httpd-2[0-9.]*
.br
uw-imap md5  ftp://ftp.cac.washington.edu/imap/ 


.B <@[A-Za-z0-9_]*@>  <string>  [<string>]

Lines starting with @ define macros for later use. 
A system macro @NAME@, which is the <name> of the URL line, can 
be used in other macro definitions.
This is handy if many URLs follow a common syntax. See
the examples for @GNU@ below. 
Macro expansion is done recursively, allowing macros to be nested. 
See the @SF@ example which works for many but not all sourceforge 
project pages.

.SH EXAMPLE

Example of a configuration file containing URL and macro
definitions:

@TAR@     @NAME@-.*?\.tar\.[bgx]z2?
.br

@SUN@     http://mirrors.sunsite.dk
.br
@GNU@     http://ftp.gnu.org/gnu/@NAME@   @TAR@
.br

@SF_RE@   /files/@NAME@/[.0-9]+|@TAR@
.br
@SF@      http://sourceforge.net/projects/@NAME@/files/@NAME@  @SF_RE@
.br

apache      md5  http://www.apache.de/dist/httpd/    httpd-2[0-9.]*
.br
exim        md5  @SUN@/exim/exim/exim4/              @TAR@
.br
gawk        md5  @GNU@
.br
hdparm      md5  @SF@
.br
uw-imap     md5  ftp://ftp.cac.washington.edu/imap/  imap-200.?[a-z]
.br

For github, gitlab, sourceforge, codeberg, and various other "forge"
sites, there's generally an Atom or RSS feed for each project. Generally,
the parts of Atom files you're interested in will be \fB<id>\fR tags,
and RSS files use \fB<link>\fR tags.

.nf
atari800    md5  https://github.com/atari800/atari800/tags.atom      <id>.*</id>
.br
sdltrs      md5  https://gitlab.com/jengun/sdltrs/-/tags?format=atom <id>.*</id>
.br
vice-emu    md5  https://sourceforge.net/projects/vice-emu/rss.xml   <link>.*</link>
.br
gumbo-parser md5  https://codeberg.org/gumbo-parser/gumbo-parser.rss <link>.*</link>
.br

.fi
Using Atom/RSS feeds saves bandwidth, vs. pulling in the full HTML
of the project's "tags" or "releases" page. For the atari800 github
project, at the time of this writing, the tags/ HTML page is 248,946
bytes, but tags.atom is only 5,267 (98% savings)!

For gitlab, github, and codeberg, the URL includes the project owner's
username. For sourceforge, the project name is enough, which means
it's easy to write a macro for those:

@SFRSS@    https://sourceforge.net/projects/@NAME@/rss.xml
.br
@L@        <link>.*</link>
.br

vice-emu     md5  @SFRSS@ @L@

.SH PROXY

If a environment variable HTTP_PROXY or http_proxy is set with a valid URI,
e.g. http://server.localnet:3128/, \fBck4up\fP make use of the proxy for
the HTTP and HTTPS protocols.

.SH FILES
\fI~/.ck4up/ck4up.conf\fP - configuration-file
.br
\fI~/.ck4up/ck4up.dbm\fP - gdbm database to store the md5 hashes

.SH AUTHOR
Juergen Daubert <jue@crux.nu>


ChangeLog
*** ck4up Revision History ***

* 1.5  Released 2026-10-03
    - New option -x to treat the args as fixed strings that must 
      match exactly one of the names in the config file.
    - New option -w prints a warning on stderr and return non-zero 
      status if a argument don't match any rule.
    - Added examples for checking github, gitlab, and codeberg 
      repos (using RSS feeds) to the man-page.
      All patches by B.Watson, many thanks.

* 1.4.1 Released 2026-01-03
    - we don't require 'resolv-replace', this breaks
      ck4up with ruby 4.0.0

* 1.4 Released 2014-12-23
    - fix redirect of http to https pages
    - update example ck4up.conf

* 1.3 Released 2014-05-08
    - follow http redirects only if response scheme is http.
      Bug reported by Daryl Fonseca-Holt.

* 1.2  Released 2009-04-14
    - Minor compatibility fixes for ruby 1.9
    - Proxy support, ck4up use the proxy server, set by the
      environment variable HTTP_PROXY, for the HTTP and HTTPS
      protocols.
      Suggested and initial patch by Ciprian Dorin.
    - Redirection isn't optional any more but used always.

* 1.1  Released 2009-02-08
    - New option -p or --parseonly to parse the configuration
      file, expand macros and print the result without fetching
      any page. Useful for debugging the config file.
    - catch recursion errors in configuration file.
    - use '/usr/bin/env ruby' as shebang line, portability fixes
      for Makefile. Suggested by Jose V Beneyto.
    - Fixed Sourceforge example in the man-page again.
    - Minor code cleanup.

* 1.0  Released 2008-04-18
    - Follow HTTP redirects, configurable via a new variable
      Follow_redirects. Enabled by default.
      Patch by Tilman Sauerbeck.
    - Enable getting pages via HTTPS, patch by Sam Hall.
    - Changed option -f or --config to accept the full path
      to the configuration file, file extension required.
      Default is still ~/.ck4up/ck4up.conf.
      Suggested by Jose V Beneyto.
    - Fixed Sourceforge example in the man-page.

* 0.2.4 Released 2007-10-25
    - Remove duplicated hits from result. Incompatible
      change, because the result differs for many sites.
      Patch by Tilman Sauerbeck
    - Use fixed user/url for ftp logins.
      Patch by Tilman Sauerbeck 
    - Bugfix: flag after precision in printf call.
    - Bugfix: don't try to write empty keys to the database 
      in function clean_db.
    - Sync stuff is not necessary, removed.

* 0.2.3 Released 2006-05-21
    - New option -f or --config to use an alternative config
      file. Without that option defaults to ck4up.conf which
      is the old behaviour.
    - Error messages are now printed to STDERR, all other to
      STDOUT.

* 0.2.2 Released 2005-10-01
    - New entries in the config file will be shown as new
      and not as diff, which was annoying. 
      Patch by Johannes Winkelmann 
    - New option -c or --cleandb to remove no longer used
      entries from the ck4up database.
      Patch by Johannes Winkelmann

* 0.2.1 Released 2005-05-08
    - Add a timeout block around the ftp related part of
      fetch_page(). Necessary if a ftp server is resolvable
      but do not respond to queries. Tunable via the global
      Ftp_time variable, default is 20 seconds.

* 0.2.0 Released 2005-02-20
    - Use ruby resolver library to get multi-threading even 
      for DNS queries. Nice speed improvement.
    - Use URI instead of regex in fetch_page to detect uri 
      classes.
    - Fix thread joining. In some rare circumstance ck4up
      doesn't finish program execution.
    - Use GetoptLong to parse commandline options. Now long
      options are available and options can be combined.
    - Join commandline arguments with '|' to build a or'ed
      regular expression.
    - Move test of Basedir into a function.
    - Use sync library instead of Thread.critical

* 0.1.10 Released 2005-02-06
    - Add option to use passive mode with ftp, which is the
      default now. Can be changed with the Ftp_passiv variable.
      Patch by Jukka Heino
    - Concat the output of the ftp list command with newlines,
      which simplifies parsing with regular expressions.
      Patch by Per Liden
    - Allow starting capital letters in the program names.
      Patch by Per Liden

* 0.1.9 Released 2004-07-28
    - raise an error if the result of a query is empty

* 0.1.8 Released 2004-06-05
    - No code changes
    - Adjust man page and sample ck4up.config for the changed 
      behaviour of the sourceforge mirrors

* 0.1.0-0.1.7
    - No change log was maintained for this releases



ck4up.rb
#! /usr/bin/env ruby
#
# ck4up
# 
# Copyright (c) Juergen Daubert <jue@crux.nu>
# Version 1.5 2026-09-29
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, 
# USA.
#

require 'net/http'
require 'net/ftp'
require 'digest'
require 'gdbm'
require 'getoptlong'
require 'timeout'


DefaultConfig = ENV['HOME'] + '/.ck4up/ck4up.conf'
HttpProxy     = ENV['HTTP_PROXY'] || ENV['http_proxy']

Threads_max   = 20    # Number of parallel threads
Ftp_passive   = true  # use passive mode for ftp
Ftp_time      = 20    # timeout in seconds for ftp requests



def usage()
	print <<EOS
Usage: ck4up [options] [exp ...]
Options: 
 -d,      --debug        debug mode, print fetched pages
 -h,      --help         print this help message
 -k,      --keep         keep md5 values, don't change database
 -c,      --cleandb      clean unused keys from database
 -p,      --parseonly    process and print configuration file
 -v,      --verbose      verbose mode, show unchanged pages
 -x,      --exact        treat arguments as exact matches
 -w,      --warn         print warning if no arguments match
 -f file, --config file  use configuration from file, see ck4up(1)
 exp                     check only configuration lines matching exp  
EOS
	exit
end


def parse_options()
	options = {}
	begin
		valid_options = GetoptLong.new(
			["--debug",     "-d", GetoptLong::NO_ARGUMENT],
			["--keep",      "-k", GetoptLong::NO_ARGUMENT],
			["--verbose",   "-v", GetoptLong::NO_ARGUMENT],
			["--cleandb",   "-c", GetoptLong::NO_ARGUMENT],
			["--help",      "-h", GetoptLong::NO_ARGUMENT],
			["--parseonly", "-p", GetoptLong::NO_ARGUMENT],
			["--exact",     "-x", GetoptLong::NO_ARGUMENT],
			["--warn",      "-w", GetoptLong::NO_ARGUMENT],
			["--config",    "-f", GetoptLong::REQUIRED_ARGUMENT])
		valid_options.each do |opt,arg|
			case opt
				when "--debug"     then options["debug"] = true
				when "--keep"      then options["keep"] = true
				when "--verbose"   then options["verbose"] = true
				when "--cleandb"   then options["cleandb"] = true
				when "--parseonly" then options["parseonly"] = true
				when "--exact"     then options["exact"] = true
				when "--warn"      then options["warn"] = true
				when "--config"    then options["config"] = arg
				when "--help"      then usage
			end
		end
	rescue
		exit -1
	end

	if ! $Config = options["config"]
		$Config = DefaultConfig
	end
	$Database = $Config.sub(/\.[^.]*$/,"") + ".dbm"

	return options
end


def print_result(io,a,b,*c)
	io.printf("%-.15s %-s %-6s %s\n", a,"."*([15-a.length,0].max),b,c.join(" "))
end


class Parser
	def initialize(file)
		@file = file
		@macro = {}
		Parser.exist_config(@file)
	end

	def Parser.exist_config(file)
		if not FileTest::exist?(file)
			puts "Error: Configuration file #{file} not found !"
			exit -1
		end
	end

	def parse
		File.read(@file).each_line do |row|
			yield replace_macros(row)
		end
	end

	def replace_macros(line)
		case line
			when /^@\w*@/
				@macro[line.split[0]] = line.split[1..-1].join(' ')
				return false
			when /^[a-zA-Z]/
				expand_macros(line)
				return line.gsub('@NAME@',line.split[0])
			else
				return false
		end
	end

	def expand_macros(line)
	 	begin
			@macro.each { |k,v| expand_macros(line) if line.gsub!(k,v) }
		rescue => error
			print_result(STDERR,line.split[0],'fatal error:',error.to_s.strip)
			exit -1
		end
	end
end


class CheckUp

	SAME = 0
	DIFF = 1
	NEW  = 2

	def CheckUp.set_http_proxy(proxy)
		@@proxy_host, @@proxy_port = nil
		return if not proxy
		p = URI.parse(proxy)
		if (p.host && p.port)
			@@proxy_host = p.host
			@@proxy_port = p.port
		else
			puts "Warning: invalid HTTP_PROXY environment variable -> " + proxy
		end
	end

	def CheckUp.set_db(db,access)
		@@db_readonly = access
		if FileTest::exist?(db)
			@@db = GDBM.open(db)
		else
			puts "Info: Creating new database #{db}"
			@@db = GDBM.new(db)
		end
		at_exit { @@db.close }
	end

	def check(name,type,url=nil,regexp=nil)
		@name = name
		@url=url
		case type
			when 'md5' then check_md5(regexp)
			else raise "Unknown type: #{type}"
		end
	end

	def check_md5(reg)
		page = fetch_page
		page = page.scan(/#{reg}/).uniq if reg
		puts page if Opts["debug"]
		page = page.join if page.is_a?(Array)
		raise "empty result" if page == ""
		save_md5(Digest::MD5.hexdigest(page))
	end

	def save_md5(md5)
		if md5 == @@db[@name]
			return SAME
		else
			@@db[@name] ? res = DIFF : res = NEW
			@@db[@name] = md5 if not @@db_readonly
			return res
		end
	end

	def clean_db(keeplist)
		active = Hash.new
		for k in keeplist
			active[k] = @@db[k]
		end

		oldCount = @@db.size
		@@db.clear

		active.each_key do |key|
			@@db[key] = active[key] if active[key]
		end

		return oldCount - @@db.size
	end

	def fetch_page()
		begin
			uri = URI.parse(@url)
			case uri.scheme
				when "http", "https"
					fetch_page_http(uri)
				when "ftp"
					fetch_page_ftp(uri)
				else
					raise "wrong url syntax #{@url}"
			end
		rescue Exception => err
			raise err.to_s
		end
	end

	def fetch_page_http(uri, limit = 10)
		if limit.zero?
			raise ArgumentError, "HTTP redirect too deep"
		end

		http = Net::HTTP.new(uri.host, uri.port, @@proxy_host, @@proxy_port)
		if uri.scheme == "https"
			http.use_ssl = true
			http.verify_mode = OpenSSL::SSL::VERIFY_NONE
		end
		response = http.request_get(uri.request_uri)

		case response
			when Net::HTTPSuccess
				return response.body
			when Net::HTTPRedirection
				loc = URI.parse(response['Location'])
				uri = loc if loc.scheme == "http" or loc.scheme == "https"
				fetch_page_http(uri, limit - 1)
			else
				response.error!
		end
	end

	def fetch_page_ftp(uri)
		Timeout::timeout(Ftp_time) do
			ftp = Net::FTP.new(uri.host)
			ftp.passive = true if Ftp_passive
			ftp.login("anonymous", "ck4up@example.com")
			ftp.chdir(uri.path)
			res = ftp.list.join("\n")
			ftp.close
			return res
		end
	end

end



def do_parse()
	Parser.new($Config).parse do |line|
		if line and line.index(/#{ARGV.join('|')}/)
			puts line
		end
	end
end

def do_cleanup()
	checkUp = CheckUp.new
	keeplist = []

	Parser.new($Config).parse do |line|
		if line
			n,t,u,r = line.split
			keeplist.push(n)
		end
	end

	count = checkUp.clean_db(keeplist)
	printf "Removed %d records\n", count
end

def do_check()
	matched = false
	threads = []

	Parser.new($Config).parse do |line|

		while Thread.list.size > Threads_max; sleep 1; end

		if line and line.index(/#{ARGV.join('|')}/)
			matched = true
			n,t,u,r = line.split
			threads << Thread.new(n,t,u,r) do |name,type,url,regexp|
				begin
					result = CheckUp.new.check(name,type,url,regexp)
					case result
						when CheckUp::SAME
							print_result(STDOUT,name,'ok') if Opts["verbose"]
						when CheckUp::NEW
							print_result(STDOUT,name,'new: ',url)
						else
							print_result(STDOUT,name,'diff:',url)
					end
				rescue => error
					print_result(STDERR,name,'error:',error.to_s.strip)
				end
			end
		end
	end

	threads.each { |t| t.join }
	return matched
end

def mangle_argv()
	ARGV.map! do |arg|
		arg = "^" + arg + "\\s"
	end
end

trap('INT') { puts; exit }

Opts = parse_options
Parser.exist_config($Config)
CheckUp.set_db($Database,Opts["keep"])
CheckUp.set_http_proxy(HttpProxy)

if Opts["cleandb"]
	do_cleanup
	exit
elsif Opts["parseonly"]
	do_parse
	exit
else
	if Opts["exact"]
		mangle_argv
	end
	if not do_check
		if Opts["warn"]
			STDERR.print("Warning: no arguments matched the config file\n")
			exit 1
		end
	end
	exit 0
end


# vim:ts=2 sw=2 noexpandtab
# End of file
